tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/publicAsset/PublicAssetRoutes.scala

```diff
@@ -0,0 +1,121 @@
+import io.netty.buffer.Unpooled
+import io.netty.handler.codec.http.HttpHeaderNames.{CACHE_CONTROL, CONTENT_LENGTH, CONTENT_TYPE}
+import io.netty.handler.codec.http.HttpResponseStatus._
+import io.netty.handler.codec.http.{HttpMethod, HttpResponseStatus}
+import jakarta.inject.Inject
+import org.apache.james.blob.api.BlobStore
+import org.apache.james.blob.api.BlobStore.StoragePolicy
+import org.apache.james.core.Username
+import org.apache.james.jmap.HttpConstants.JSON_CONTENT_TYPE
+import org.apache.james.jmap.api.model.Size.Size
+import org.apache.james.jmap.core.ProblemDetails
+import org.apache.james.jmap.json.ResponseSerializer
+import org.apache.james.jmap.{Endpoint, JMAPRoute, JMAPRoutes}
+import org.apache.james.mailbox.model._
+import org.apache.james.util.ReactorUtils
+import org.slf4j.{Logger, LoggerFactory}
+import play.api.libs.json.Json
+import reactor.core.publisher.Mono
+import reactor.core.scala.publisher.SMono
+import reactor.core.scheduler.Schedulers
+import reactor.netty.http.server.{HttpServerRequest, HttpServerResponse}
+
+import scala.compat.java8.FunctionConverters._
+import scala.util.{Success, Try}
+
+object PublicAssetRoutes {
+  val LOGGER: Logger = LoggerFactory.getLogger(classOf[PublicAssetRoutes])
+  val ONE_YEAR_AS_SECONDS: Int = 31536000
+  val BUFFER_SIZE: Int = 16 * 1024
+}
+
+class PublicAssetRoutes @Inject()(val publicAssetRepository: PublicAssetRepository,
+                                  val blobStore: BlobStore) extends JMAPRoutes {
+
+  private val usernameParam: String = "username"
+  private val assetIdParam: String = "assetId"
+  private val publicAssetUri = s"/publicAsset/{$usernameParam}/{$assetIdParam}"
+
+  override def routes(): stream.Stream[JMAPRoute] = Stream.of(
+    JMAPRoute.builder
+      .endpoint(new Endpoint(HttpMethod.GET, publicAssetUri))
+      .action(this.getAsset)
+      .corsHeaders,
+    JMAPRoute.builder
+      .endpoint(new Endpoint(HttpMethod.OPTIONS, publicAssetUri))
+      .action(JMAPRoutes.CORS_CONTROL)
+      .noCorsHeaders)
+
+  private def getAsset(request: HttpServerRequest, response: HttpServerResponse): Mono[Void] = {
+    val usernameTry: Try[Username] = Try(Username.of(request.param(usernameParam)))
+    val assetIdTry: Try[PublicAssetId] = PublicAssetId.fromString(request.param(assetIdParam))
+
+    (usernameTry, assetIdTry) match {
+      case (Success(username), Success(assetId)) => getAsset(request, response, username, assetId)
+      case _ => returnNotFound(response)
+          .asJava()
+          .`then`
+    }
+  }
+
+  private def getAsset(request: HttpServerRequest, response: HttpServerResponse, username: Username, assetId: PublicAssetId): Mono[Void] =
+    SMono(publicAssetRepository.get(username, assetId))
+      .switchIfEmpty(SMono.error(PublicAssetNotFoundException(assetId)))
+      .flatMap(asset => SMono(blobStore.readReactive(blobStore.getDefaultBucketName, asset.blobId, StoragePolicy.LOW_COST))
```

Comment:
Why does the blobstore, a storage detail, leak into the JMAP presentation layer?

I would expect the repository to encapsulate such details...

Reply:
Should not actually (I relied on the old repository API).